src/vdaf/prg.rs

```diff
@@ -0,0 +1,232 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module implements PRGs as specified in draft-patton-cfrg-vdaf-01.
+
+use crate::vdaf::{CodecError, Decode, Encode};
+use aes::{
+    cipher::{
+        generic_array::GenericArray, FromBlockCipher, NewBlockCipher,
+        StreamCipher as AesStreamCipher,
+    },
+    Aes128, Aes128Ctr,
+};
+use cmac::{Cmac, Mac, NewMac};
+use std::{
+    fmt::Debug,
+    io::{Cursor, Read},
+};
+
+/// Input of [`Prg`].
+#[derive(Clone, Debug, Eq)]
+pub struct Seed<const L: usize>(pub(crate) [u8; L]);
+
+impl<const L: usize> Seed<L> {
+    /// Generate a uniform random seed.
+    pub fn generate() -> Result<Self, getrandom::Error> {
+        let mut seed = [0; L];
+        getrandom::getrandom(&mut seed)?;
+        Ok(Self(seed))
+    }
+
+    pub(crate) fn uninitialized() -> Self {
+        Self([0; L])
+    }
+
+    pub(crate) fn xor_accumulate(&mut self, other: &Self) {
+        for i in 0..L {
+            self.0[i] ^= other.0[i]
+        }
+    }
+
+    pub(crate) fn xor(&mut self, left: &Self, right: &Self) {
+        for i in 0..L {
+            self.0[i] = left.0[i] ^ right.0[i]
+        }
+    }
+}
+
+impl<const L: usize> PartialEq for Seed<L> {
+    fn eq(&self, other: &Self) -> bool {
+        // Do constant-time compare.
+        let mut r = 0;
+        for (x, y) in (&self.0[..]).iter().zip(&other.0[..]) {
+            r |= x ^ y;
+        }
+        r == 0
+    }
+}
+
+impl<const L: usize> Encode for Seed<L> {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        bytes.extend_from_slice(&self.0[..]);
+    }
+}
+
+impl<const L: usize> Decode<()> for Seed<L> {
+    fn decode(_decoding_parameter: &(), bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let mut seed = [0; L];
+        bytes.read_exact(&mut seed)?;
+        Ok(Seed(seed))
+    }
+}
+
+/// A stream of pseudorandom bytes derived from a seed.
+pub trait SeedStream: Sized {
```

Comment:
Is this `Sized` requirement because of the methods on `trait Prg` that return `Self::SeedStream`? I think it could be argued that the `Sized` bound then belongs in `Prg` (i.e., it should say `type SeedStream: SeedStream + Sized;`) but if we don't expect `SeedStream` to ever be used independently from `Prg` then this is fine.

Reply:
This turned out to not be necessary, so I removed it.